### ISSUE TITLE
proposal: quarantine flaky tests

### ISSUE THREAD

**Issue by andrewhsu** [moby/moby#39436]
Talked with @kolyshkin @cpuguy83 @tiborvass @thaJeztah about this...

Scope of this issue is just the `test-integration` tests that run a test with cli against a daemon for each PR. Unit tests are rock-solid dependable.

Looks like [only 7 of the last 25 PRs merged (28%)](https://github.com/moby/moby/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Amerged+closed%3A%3C2019-06-28) have all PR checks pass green, but they are merged anyways mostly because a human manually determines the failed tests are flaky. I'm assuming the current criteria of merging PRs with flaky tests is acceptable because of a limited amount of consistently flaky `test-integration` tests, approximately 3.3% based on [identified list by github issue](https://github.com/moby/moby/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+flaky).

A typical cycle I've observed is:

1. PR is opened, quorum of LGTMs reached.
2. Almost 3 hrs later, after all tests have run, reviewer sees PR check is red and sees one (or several) of the `test-integration` tests are red even though the PR does not change the codebase related to the failed test. Reviewer verifies failed test is one of [approximately 50 known flaky tests](https://github.com/moby/moby/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+flaky) out of the suite of approximately 1500 `test-integration` tests run and kicks the PR job again to see if the flaky test passes without any additional code changes.
3. Reviewer watches a [Marvel movie](https://www.imdb.com/title/tt4154796/) (or any equally entertaining movie for one's taste), returns to PR checks after 3 hrs when all tests have completed a second run to see original flaky test passed, but a new flaky test failed.
4. Reviewer digs into flaky test a bit more, asks around, maybe kicks PR checks again to see if everything goes green, debates re-running PR jobs again if same result, reconsiders, and then decides to merge with red PR.

## proposal 1

As an improvement to this process (because it seems to be an acceptable process at this moment in time), I'd like to propose if a flaky test is encountered from now on:

1. Identify the `test-integration` test as flaky in the codebase some way (perhaps renaming the test to begin with `TestFlaky`).
2. Have the PR check automatically rerun each flaky test up to 5 times and if the flaky test passes at least once, then mark the flaky test as successful. i.e. have 20% pass rate as the acceptable bar for a test marked as flaky.
3. Have a separate report published somewhere at the end of the PR check that summarizes:
  a) name of each flaky test
  b) how many times each flaky test was run
  c) if the flaky test passed
  d) how many flaky tests were run in total

This will help in these ways:

1. Developer can identify flaky tests in code as opposed to searching in github issues. This helps correlate when a test is first deemed flaky, to compare against when the test was written.
2. Saves time on the process of merging a PR by decreasing the time from PR mergeable after quorum of LGTMs to actual merge. This is a savings of approximately 3 hrs per re-run of the `test-integration` tests.
3. Higher percentage of PRs merged have green status, i.e. higher than 28% at this moment, without sacrificing our criteria for how we currently override a flaky test status. If we attach an [auto merge bot](https://github.com/moby/moby/issues/35490#issuecomment-364296583) a higher percentage of green tests can help get more PRs merged and reduce time from PR open to PR merge.
4. The flaky test report will help track if a flaky test is getting more flaky or less flaky which can drive attention towards how much "worse" a test is getting.

The proposal of 5 times run is arbitrary. Not sure if we need more analysis to see if that number is too high. I would not like to propose we have the number increased beyond that otherwise my gut says less than 20% pass rate is not a helpful test.

Having an automated rerun process acknowledges that flaky tests are still valuable in some way in providing information about the quality of the proposed code changes in a PR, e.g. if a flaky test usually passes 50% of the time, but then consistently fails with a new PR then the developer can use that as a signal.

Of course I'd prefer to have our PR checks addressed in a manner that we can solidly depend on them to accurately identify undesirable code in a PR without flakiness, i.e. fix all the flaky tests. I think we can work on that in parallel.

## proposal 2

Same as proposal 1, but have the flaky tests rerun in parallel to the non-flaky tests so as to not take up any extra time. Also, the results of the rerun of the flaky tests will not block the PR checks from going green if any of the flaky tests fail after 5 runs, i.e. flaky tests always treated as green result even if the test fails. The flaky test reports are still published and can be used to help measure the varying degrees of flakiness changes assuming this is still valuable information to gather.

## proposal 3

Same as proposal 1, but do not run any tests marked as flaky. This also means a separate report on the results of the flaky tests do not need to be published because there would be no results.

**Comment by crosbymichael** [comment]
If a test is flaky it is a bad test.  Remove it.  That whole run things 5 times just adds time to our tests and makes them even more frustrating and why I sit for hours, 4-6hrs, waiting for one line changes in docker to eventually fail because of randomness.

The tests should pass 100% of the time unless there is a bug introduced in a code change.  Anything else is just a worthless way to run a CI. 

I vote to remove these flaky tests, remove the run 5 times to fine flaky tests as this is just accepting we cannot write solid tests, and then either keep these bad tests out or add them back when someone knows how to write a solid test.

**Comment by crosbymichael** [comment]
The CI is embarrassing right now

**Comment by andrewhsu** [comment]
@crosbymichael I've created 3 proposals with my original proposal as `proposal 1` and I believe your suggestion is effectively the `proposal 3` derivative.

I've also created a `proposal 2` as a middle-ground option.

**Comment by cpuguy83** [comment]
If we do 2 or 3 we might as well remove the tests.

I'm hesitant to blindly remove the tests as a whole just because there is *some* value in that if a test goes from flaky to absolutely failing then something has gone wrong. Of course this is hard to track since it's mostly just on maintainers to keep track of what failed when and questioning things like "I *feel* like this went from mostly working to... less than mostly working"

However the current strategy just doesn't work and wastes a *lot* of CPU time.

**Comment by seemethere** [comment]
If we have a full list of current flaky tests we should focus on the functionality they're trying to solve for and rewrite them in a way that is valuable.

I agree with @crosbymichael if a test is flaky it is a bad test.

**Comment by kolyshkin** [comment]
I tried to figure out failure causes in some swarm tests (in order to fix those), spent some considerable time on that and unfortunately got nowhere. I wish someone else could attack these "flakes", maybe they'll be more successful at it than I was.

Maybe we need to be **bold** here and just drop the flaky tests (even without re-implementing them). Currently, they don't add any value (all we do is say "a known flay test" and move on), but are very expensive, in terms of precious developers/maintainers time spend looking at these failures.

**Comment by mavenugo** [comment]
👍 to @crosbymichael suggestion. Let us open an issue to track each flaky test being removed out of test cycle. It will be great to have a green run and focus on re-adding the flaky ones after fixing the tests or test infra appropriately.

**Comment by thaJeztah** [comment]
@mavenugo we're looking at a possible root-cause for all the failures; basically comes down to _all_ SwarmKit tests would be flaky; https://github.com/moby/moby/issues/34051#issuecomment-511073471

If multiple daemons are started (or the daemon of a previous test did not yet fully clean up), tests might fail because we're always using the same port; https://github.com/moby/moby/blob/ec86547244fa329148a096db56f9ade77a7ce7eb/vendor/github.com/hashicorp/memberlist/config.go#L226-L231

@ehazlett found that in libnetwork, the problem is circumvented by incrementing the port for each test; https://github.com/docker/libnetwork/blob/67b7970d93313b2c6032e4e8742baec92fabbdb3/networkdb/networkdb_test.go#L42-L45

We'll probably need something similar in the tests in this repository

**Comment by andrewhsu** [comment]
For the initial scope of this issue, I'm withdrawing my proposal to quarantine flaky tests because it looks like we've come a long way in terms of stabilizing the checks to the point where we can get a useful signal.

When this issue was open, we had 32% green of the [25 recently-merged](https://github.com/moby/moby/pulls?utf8=%E2%9C%93&q=is%3Apr+merged%3A%3C%3D2019-10-02).

Now we have 92% green of the [25 recently-merged](https://github.com/moby/moby/pulls?utf8=%E2%9C%93&q=is%3Apr+merged%3A%3C%3D2019-10-02
).
